include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+                        const bool                                                 enforce_zero_boundary,
+                        const Quadrature<dim-1>                                   &q_boundary,
+                        const bool                                                 project_to_boundary_first)
+  {
+    const parallel::Triangulation<dim,spacedim> *parallel_tria =
+      dynamic_cast<const parallel::Triangulation<dim,spacedim>*>(&dof.get_tria());
+    Assert (project_to_boundary_first == false, ExcNotImplemented());
+    Assert (enforce_zero_boundary == false, ExcNotImplemented());
+    (void) enforce_zero_boundary;
+    (void) project_to_boundary_first;
+    (void) q_boundary;
+
+    const IndexSet locally_owned_dofs = dof.locally_owned_dofs();
+    IndexSet locally_relevant_dofs;
+    DoFTools::extract_locally_relevant_dofs(dof, locally_relevant_dofs);
+
+    typedef typename VectorType::value_type number;
+    Assert (dof.get_fe().n_components() == function.n_components,
+            ExcDimensionMismatch(dof.get_fe().n_components(),
+                                 function.n_components));
+    Assert (vec_result.size() == dof.n_dofs(),
+            ExcDimensionMismatch (vec_result.size(), dof.n_dofs()));
+    Assert (dof.get_fe().degree == fe_degree,
+            ExcDimensionMismatch(fe_degree, dof.get_fe().degree));
+    Assert (dof.get_fe().n_components() == components,
+            ExcDimensionMismatch(components, dof.get_fe().n_components()));
+
+    // set up mass matrix and right hand side
+    typename MatrixFree<dim,number>::AdditionalData additional_data;
+    additional_data.tasks_parallel_scheme =
+      MatrixFree<dim,number>::AdditionalData::partition_color;
+    if (parallel_tria)
+      {
+        const MPI_Comm &mpi_communicator = parallel_tria->get_communicator();
+        additional_data.mpi_communicator = mpi_communicator;
+      }
+    additional_data.mapping_update_flags = (update_values | update_JxW_values);
+    MatrixFree<dim, number> matrix_free;
+    matrix_free.reinit (mapping, dof, constraints,
+                        QGauss<1>(fe_degree+2), additional_data);
+    typedef MatrixFreeOperators::MassOperator<dim, fe_degree, components, number> MatrixType;
+    MatrixType mass_matrix;
+    mass_matrix.initialize(matrix_free);
+    mass_matrix.compute_diagonal();
+
+    typedef LinearAlgebra::distributed::Vector<number> LocalVectorType;
+    LocalVectorType vec, rhs, inhomogeneities;
+    matrix_free.initialize_dof_vector(vec);
+    matrix_free.initialize_dof_vector(rhs);
+    matrix_free.initialize_dof_vector(inhomogeneities);
+    constraints.distribute(inhomogeneities);
+    inhomogeneities*=-1.;
+
+    create_right_hand_side (mapping, dof, quadrature, function, rhs, constraints);
+    mass_matrix.vmult_add(rhs, inhomogeneities);
+
+    //now invert the matrix
+    ReductionControl     control(5*rhs.size(), 0., 1e-12, false, false);
```

Comment:
That's quite bad. What is the polynomial degree in `FE_Q_Hierarchical`. I can imagine that happening if the polynomial degree gets four or higher because `FE_Q_Hierarchical` is really badly conditioned at higher degrees where roundoff forces more iterations.

In any case, I definitely agree that we should stay at `5*rhs.size()` then. It's not that the usual case with well-behaved mass matrices get slower...